keps/sig-api-machinery/2839-in-use-protection/README.md

```diff
@@ -0,0 +1,652 @@
+However, users may think that the successful addition of the `Liens` means that the resource won't be deleted until the `Liens` are deleted, which isn't true.
+To avoid such a misunderstanding from users, a request for adding additional `Liens` to a resource with non-nil `DeletionTimestamp` should be blocked by validating admission webhook for update.
+
+#### Cross Namespace Dependency Consideration
+
+`Liens` can be added only by users or controllers that have create/update rights to resources.
+Therefore, this mechanism itself doesn't have issues for cross namespace dependency.
+However, when it is used with controllers that have access across namespaces,
+they can easily introduce a cross namespace dependency.
+For example, adding `Liens` to resource X in namespace x, because resource Y exists in namespace y.
+This may allow malicious users to block the resources in other namespace from being deleted.
+
+To minimize issues caused by a cross namespace dependency, it should be considered to implement
+a mechanism to opt-in/opt-out adding liens across namespace in such controllers.
+For example, a controller may choose to add `Liens` only to resources that has `example.com/sample-controller/allow-cross-namespace-liens: true` annotation, if the dependency resource isn't in the same namespace.
+Such annotaions can be an allow list and/or a deny list for namespaces instead of boolean flag,
+depending on use cases, however it's up to consumers to decide.
+
+#### Per Controller `Liens` vs Per Reason `Liens`
+
+The same controllers may need to add `Liens` to a certain resource for multiple reasons, like multiple resources are using it.
+In such a case, if `Liens` are added per controller basis, such a controller needs to be sure that no other resources are actually using the resource, when removing its `Liens` from the resource.
+On the other hand, if `Liens` are added per reason basis, such a controller can remove the `Lien` for the resource, without considering other resources.
+Also, the latter way helps users to know which particular resources are using the resource.
+
+`Liens` allows both per controller `Liens` and per reason `Liens`.
+Only requirements are to include identity that who add it and to only remove its own's.
+
+Note that examples of per controller `Liens` and per reason `Liens` are as follows:
+
+Let's assume that a secret protection controller would like to protect Secret `my-secret` in `ns1` when
+both PersistentVolume `my-pv1` and Pod `my-pod1` in namespace `ns2` are using it.
+
+- Per Controller `Liens`:
+```
+apiVersion: v1
+kind: Secret
+metadata:
+  namespace: ns1
+  name: my-secret
+  liens:
+    - "k8s.io/secret-protection-controller"
+```
+
+- Per Reason `Liens`:
+```
+apiVersion: v1
+kind: Secret
+metadata:
+  namespace: ns1
+  name: my-secret
+  liens:
+    - "k8s.io/secret-protection-controller v1/PersistentVolume//my-pv1"
+    - "k8s.io/secret-protection-controller v1/Pod/ns2/my-pod1"
```

Comment:
k/v pairs would make this much prettier. Or a list of structs with {owner, reason} fields. etc.